src/frontend/src/optimizer/plan_node/batch_hash_agg.rs

```diff
@@ -56,6 +57,43 @@ impl BatchHashAgg {
     pub fn group_key(&self) -> &[usize] {
         self.logical.group_key()
     }
+
+    fn to_two_phase_agg(&self, dist_input: PlanRef) -> Result<PlanRef> {
+        // partial agg - follows input distribution
+        let partial_agg: PlanRef = self.clone_with_input(dist_input).into();
+
+        // insert exchange
+        let exchange = RequiredDist::shard_by_key(
+            partial_agg.schema().len(),
+            &(0..self.group_key().len()).collect_vec(),
+        )
+        .enforce_if_not_satisfies(partial_agg, &Order::any())?;
```

Comment:
Since this method is intended to be called only when `hash_agg_dist_satisfied_by_input_dist` is false, is it better to `debug_assert` the type of `PlanNode` here? Just in case of misuse.